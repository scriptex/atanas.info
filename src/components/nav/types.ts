import type { MenuItem } from '@data/menu';
import type { ReactChildren } from '@scripts/types';

export type Props = {
	onClick?: () => void;
	hasShell?: boolean;
	children?: ReactChildren;
	className?: string;
};

export type AnchorProps = Omit<MenuItem, 'children'> & Pick<Props, 'onClick'>;
