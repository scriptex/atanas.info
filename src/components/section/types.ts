import type { CSSProperties, ReactNode } from 'react';

import type { ReactChildren } from '@scripts/types';

export type Props = {
	actions?: ReactNode;
	additionalElements?: ReactNode;
	children: ReactChildren;
	className?: string;
	hasButton?: boolean;
	hasShell?: boolean;
	id: string;
	shellClass?: string;
	style?: CSSProperties;
	subtitle?: string;
	title?: string;
};
