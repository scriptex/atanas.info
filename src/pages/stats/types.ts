export type Package = {
	name: string;
	version: string;
	description: string;
	license: string;
	homepage: string;
	author: string;
	downloads: number;
};

export type Packages<T = Record<string, Package>> = {
	data: Record<string, Package> & T;
};

export type WithSum = {
	sum: number;
};

export type WithError = {
	error?: boolean;
};

export type Props = Packages<WithSum & WithError>;
