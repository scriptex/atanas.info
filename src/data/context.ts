import { createContext } from 'react';

export const AppContext = createContext({
	contactVisible: false,
	setContactVisible: (state: boolean) => state
});
