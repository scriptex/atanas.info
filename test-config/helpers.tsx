import { act } from 'react-dom/test-utils';
import type { FC } from 'react';
import { render, RenderResult, waitFor } from '@testing-library/react';

import * as shared from '@scripts/shared';

type ComponentType = FC<Readonly<React.ComponentProps<any>>>;

export const test = async (Component: ComponentType): Promise<RenderResult> => {
	let result!: RenderResult;

	await act(async () => {
		result = await waitFor(() => render(<Component />));
	});

	return result;
};

export const snapshotTest = (Component: ComponentType, buttonSelector?: string): void => {
	const { name } = Component;

	jest.spyOn(shared, 'formatDate');

	(shared.formatDate as jest.Mock).mockImplementation(() => 'Mock date');

	afterAll(() => {
		jest.clearAllMocks();
	});

	describe(`${name} component`, () => {
		it(`Should render the ${name} component`, async () => {
			const { asFragment, container } = await test(Component);

			expect(asFragment()).toMatchSnapshot();

			if (buttonSelector) {
				await act(async () => {
					await container.querySelectorAll<HTMLElement>(buttonSelector).forEach(button => {
						if (!button.hasAttribute('download')) {
							button.click();

							expect(asFragment()).toMatchSnapshot();
						}
					});
				});
			}
		});

		it(`Should unmount the ${name} component`, async () => {
			const { unmount, asFragment } = await test(Component);

			unmount();

			expect(asFragment()).toMatchSnapshot();
		});
	});
};

export const mockAudioContext = (): jest.Mock => {
	const mockConnect = jest.fn();

	const mockgetByteFrequencyData = jest.fn();

	const mockcreateMediaElementSource = jest.fn(() => ({
		connect: mockConnect
	}));

	const mockcreateAnalyser = jest.fn(() => ({
		connect: mockConnect,
		frequencyBinCount: [0, 1, 2],
		getByteFrequencyData: mockgetByteFrequencyData
	}));

	const mockcreateOscillator = jest.fn(() => ({
		channelCount: 2
	}));

	const mockChannelSplitterConnect = jest.fn(n => n);

	const mockcreateChannelSplitter = jest.fn(() => ({
		connect: mockChannelSplitterConnect
	}));

	const mockedAudioContext = jest.fn(() => ({
		createAnalyser: mockcreateAnalyser,
		createOscillator: mockcreateOscillator,
		createChannelSplitter: mockcreateChannelSplitter,
		createMediaElementSource: mockcreateMediaElementSource
	}));

	return mockedAudioContext;
};

export const mockFetch = (value: Record<string, string | boolean>): void => {
	global.fetch = jest.fn().mockImplementation(() =>
		Promise.resolve({
			json: () => Promise.resolve(value)
		})
	);
};
