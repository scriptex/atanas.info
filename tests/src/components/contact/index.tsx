import { act } from 'react';

import { Contact } from '@components';
import { AppContext } from '@data/context';
import { Status } from '@scripts/types';
import { mockFetch, snapshotTest, test } from '@test-config/helpers';

const fetch = global.fetch;
const context = { contactVisible: true, setContactVisible: jest.fn() };

afterAll(() => {
	global.fetch = fetch;
	jest.clearAllMocks();
});

snapshotTest(Contact, '.c-contact__close');
snapshotTest(() => (
	<AppContext.Provider value={context}>
		<Contact />
	</AppContext.Provider>
));
snapshotTest(() => <Contact initialStatus={Status.ERROR} />);
snapshotTest(() => <Contact initialStatus={Status.SUCCESS} />);

it('Tests Contact submission', async () => {
	const { asFragment, container } = await test(Contact);

	expect(asFragment()).toMatchSnapshot();

	const form = container.querySelector('form');

	if (!form) {
		return;
	}

	mockFetch({ error: 'Something went wrong!' });

	await act(async () => await form.submit()); // NOSONAR

	expect(asFragment()).toMatchSnapshot();

	mockFetch({ success: true });

	await act(async () => await form.submit()); // NOSONAR

	expect(asFragment()).toMatchSnapshot();
});
