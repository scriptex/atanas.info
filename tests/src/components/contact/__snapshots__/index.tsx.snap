// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` component Should render the  component 1`] = `
<DocumentFragment>
  <form
    class="c-contact c-contact--visible"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[` component Should render the  component 2`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
        <p
          class="c-contact__error"
        />
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[` component Should render the  component 3`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <div
        class="c-contact__actions"
      >
        <p
          class="c-contact__success"
        >
          Thank you for your message! I will get in touch with you soon.
        </p>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[` component Should unmount the  component 1`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 2`] = `<DocumentFragment />`;

exports[` component Should unmount the  component 3`] = `<DocumentFragment />`;

exports[`Contact component Should render the Contact component 1`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Contact component Should render the Contact component 2`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Contact component Should unmount the Contact component 1`] = `<DocumentFragment />`;

exports[`Tests Contact submission 1`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Tests Contact submission 2`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <h2>
        Send me your message
      </h2>
      <div
        class="c-contact__field"
      >
        <label
          for="email"
        >
          Email:
        </label>
        <input
          id="email"
          name="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="c-contact__field"
      >
        <label
          for="message"
        >
          Message:
        </label>
        <textarea
          cols="30"
          id="message"
          name="message"
          required=""
          rows="10"
        />
      </div>
      <input
        name="field"
        type="hidden"
        value=""
      />
      <div
        class="c-contact__actions"
      >
        <button
          class="css-8n53dr-Button"
          type="submit"
        >
          Send message
        </button>
        <p
          class="c-contact__error"
        >
          Something went wrong. Please try again.
        </p>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`Tests Contact submission 3`] = `
<DocumentFragment>
  <form
    class="c-contact"
  >
    <button
      class="c-contact__close"
      type="button"
    >
      Close
    </button>
    <div
      class="c-contact__body"
    >
      <div
        class="c-contact__actions"
      >
        <p
          class="c-contact__success"
        >
          Thank you for your message! I will get in touch with you soon.
        </p>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
